src/Chunk.ts

```diff
@@ -168,235 +148,205 @@ export default class Chunk {
+				}
 			}
-			let tracedName: string;
-			if (traced.module.chunk === this || traced.module.isExternal) {
-				tracedName = traced.name;
-			} else {
-				// if we exposed an export in another module ensure it is exported there
-				tracedName = (<Module>traced.module).chunk.ensureExport(
-					traced.module,
-					variable,
-					traced.name
-				);
+		} else {
+			// shortcut cross-chunk relations can be added by traceExport
+			for (let module of Array.from(this.imports.values())) {
+				const chunkOrExternal = module instanceof Module ? module.chunk : module;
+				if (this.dependencies.indexOf(chunkOrExternal) === -1) {
+					this.dependencies.push(chunkOrExternal);
+				}
 			}
-			this.exports[exportName] = {
-				module: traced.module,
-				name: tracedName,
-				variable
-			};
-			this.exportedVariableNames.set(variable, exportName);
-		});
+		}
 	}
 
-	collectDependencies(entryFacade?: Module) {
-		if (entryFacade) {
-			this.dependencies = [entryFacade.chunk];
-			this.entryModule = entryFacade;
-			this.isEntryModuleFacade = true;
-			return;
+	// note we assume the facade module chunk is itself linked
+	// with generateEntryExports called
+	linkFacade(entryFacade: Module) {
+		this.dependencies = [entryFacade.chunk];
+		this.entryModule = entryFacade;
+		this.isEntryModuleFacade = true;
+		for (let exportName of entryFacade.getAllExports()) {
+			const tracedVariable = entryFacade.trace(exportName);
+			this.exports.set(tracedVariable, entryFacade);
+			this.exportNames[exportName] = tracedVariable;
 		}
+	}
 
+	link() {
 		this.dependencies = [];
-
-		this.orderedModules.forEach(module => {
-			module.dependencies.forEach(dep => {
+		for (let module of this.orderedModules) {
+			for (let dep of module.dependencies) {
 				if (dep.chunk === this) {
-					return;
+					continue;
 				}
-
 				let depModule: Chunk | ExternalModule;
 				if (dep instanceof Module) {
 					depModule = dep.chunk;
 				} else {
 					// unused pure external modules can be skipped
 					if (!dep.used && this.graph.isPureExternalModule(dep.id)) {
-						return;
+						continue;
 					}
 					depModule = dep;
 				}
-
-				if (!this.dependencies.some(dep => dep === depModule)) {
+				if (this.dependencies.indexOf(depModule) === -1) {
 					this.dependencies.push(depModule);
 				}
-			});
-		});
-
-		Object.keys(this.exports).forEach(exportName => {
-			const expt = this.exports[exportName];
-			if (expt.module instanceof ExternalModule) {
-				if (!this.dependencies.some(dep => dep === expt.module)) {
-					this.dependencies.push(expt.module);
-				}
-			} else if (expt.module.chunk !== this) {
-				if (!this.dependencies.some(dep => dep === expt.module.chunk)) {
-					this.dependencies.push(expt.module.chunk);
-				}
 			}
-		});
-	}
-
-	generateImports() {
-		this.orderedModules.forEach(module => {
-			Object.keys(module.imports).forEach(importName => {
+			for (let importName of Object.keys(module.imports)) {
```

Comment:
Agreed const might be a good pattern to follow in general. Another optimization for these can be to hoist the let statement to above the for loop to avoid making a new separate variable allocation each loop iteration.